jdbc-driver/src/main/java/com/google/cloud/dataproc/jdbc/DataprocDriverUtils.java

```diff
@@ -0,0 +1,246 @@
+        abstract String host();
+
+        abstract String httpPath();
+
+        abstract String transportMode();
+
+        abstract int port();
+
+        abstract String dbName();
+
+        //  Do not parse these other parameters and pass them directly to HiveConnection
+        @Nullable
+        abstract String otherSessionConfs();
+
+        @Nullable
+        abstract String hiveConfs();
+
+        @Nullable
+        abstract String hiveVars();
+
+        static Builder builder() {
+            // Set default values
+            return new AutoValue_DataprocDriverUtils_HiveJdbcConnectionParam.Builder()
+                    .setHttpPath(HTTP_PATH)
+                    .setTransportMode(TRANSPORT_MODE)
+                    .setPort(DEFAULT_PORT)
+                    .setDbName(DEFAULT_DATABASE)
+                    .setOtherSessionConfs(null)
+                    .setHiveVars(null)
+                    .setHiveConfs(null);
+        }
+
+        public String toHiveJdbcUrl() {
+            String hiveJdbcURL =
+                    String.format(
+                            "%s%s:%d/%s;transportMode=%s;httpPath=%s",
+                            HIVE_PROTOCOL, host(), port(), dbName(), transportMode(), httpPath());
+            if (otherSessionConfs() != null)
+                hiveJdbcURL = String.format("%s%s", hiveJdbcURL, otherSessionConfs());
+            if (hiveConfs() != null) hiveJdbcURL = String.format("%s?%s", hiveJdbcURL, hiveConfs());
+            if (hiveVars() != null) hiveJdbcURL = String.format("%s#%s", hiveJdbcURL, hiveVars());
+            return hiveJdbcURL;
+        }
+
+        @AutoValue.Builder
+        abstract static class Builder {
+            abstract Builder setHost(String value);
+
+            abstract Builder setPort(int value);
+
+            abstract Builder setDbName(String value);
+
+            abstract Builder setHttpPath(String value);
+
+            abstract Builder setTransportMode(String value);
+
+            abstract Builder setOtherSessionConfs(String value);
+
+            abstract Builder setHiveConfs(String value);
+
+            abstract Builder setHiveVars(String value);
+
+            abstract HiveJdbcConnectionParam build();
+        }
+    }
+
+    /**
+     * Parses client url and extracts Hive connection parameters.
+     *
+     * @param url possible formats:
+     *     jdbc:dataproc://hive/<dbName>;clusterName=<>;projectId=<>;region=<>;port=<>
+     *     jdbc:dataproc://hive/<dbName>;clusterName=<>;projectId=<>;region=<>;clusterPoolLabel=foo=bar:env=staging;
+     * @return HiveJdbcConnectionParam object with fields correctly set
+     * @throws IOException
+     * @throws InvalidURLException
+     */
+    public static HiveJdbcConnectionParam parseHiveUrl(String url) throws InvalidURLException {
+        HiveJdbcConnectionParam.Builder paramBuilder = HiveJdbcConnectionParam.builder();
+        String urlStripped = url.substring(URL_HIVE_PREFIX.length());
```

Comment:
Similarly here and below it might be helpful to do use `URL` as a way of parsing the jdbc connection string.

Then you could do something like:

```
String path = new URL(...).getPath();
List<String> params = path.split(";");
String dbName = params.get(0); // possibly "" if no db specified
for (String param : params) {
  key, value = param.split("=")
  ...
}
```

You can also use `getQuery()` and `getRef()` for stuff after `?` and `#` respectively